src/log.c

```diff
@@ -36,70 +36,378 @@
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[EP_CM] = "ep_cm",
+	[EP_DM] = "ep_dm",
+	[AV] = "av",
+	[CQ] = "cq",
+	[EQ] = "eq",
+	[MR] = "mr"
+};
+
+const char *log_levels[] = {
+	[LOG_WARN] = "warn",
+	[LOG_TRACE] = "trace",
+	[LOG_INFO] = "info"
+};
+
+/*
+ * By default support logging from all providers and subsystems. This will
+ * change if the relevant environment variables are set.
+ */
+uint64_t log_mask = PROV_MASK | SUBSYS_MASK;
+
+typedef void found_func_t(char *ptr);
+
 /* General implementation note: these functions currently use multiple fprintfs
  * in a row, which can render in an ugly fashion for multithreaded code and for
  * some mpirun implementations.  If this bugs anyone enough then we can convert
  * them to snprintf to build up the printout in a single buffer.
  */
 
-int fi_log_level = INT_MIN;
+static int compare_names(struct slist_entry *item, const void *arg)
+{
+	struct provider_parameter *target =
+	    container_of(item, struct provider_parameter, entry);
+	return !strcmp(target->name, arg);
+}
+
+static bool check_negated(char **str)
+{
+	bool negated = false;
+
+	if (**str == '^') {
+		(*str)++;
+		negated = true;
+	}
+
+	return negated;
+}
+
+static char *alloc_env(char *name)
+{
+	char *env = getenv(name);
+	char *ret = NULL;
+
+	if (!env) {
+		return ret;
+	}
+
+	ret = strdup(env);
+	if (!ret) {
+		fprintf(stderr, "%s: %s", __func__, strerror(errno));
+	}
+
+	return ret;
+}
+
+struct prov_handle *fi_log_register_provider(const char *prov_name)
+{
+	struct prov_handle *handle;
+	struct slist_entry *found;
+	struct provider_parameter *item;
+
+	static char curr_pos = 1;
+
+	handle = calloc(sizeof(*handle), 1);
+	if (!handle) {
+		goto err;
+	}
+
+	handle->name = strdup(prov_name);
+	if (!handle->name) {
+		goto cleanup;
+	}
+
+	handle->bitpos = curr_pos;
+
+	curr_pos <<= 1;
+
+	found = slist_remove_first_match(&provider_list.names, compare_names,
+					 handle->name);
+
+	/*
+	 * Since providers are loaded dynamically we can't just negate the mask
+	 * at the end after all providers in the negated list have been added.
+	 * Need to maintain a list of providers that can print at any given
+	 * time before all of the providers have loaded.
+	 */
+	if (found) {
+		if (provider_list.negated) {
+			log_mask &= ~EXPAND(handle->bitpos, PROV_OFFSET);
+		} else {
+			log_mask |= EXPAND(handle->bitpos, PROV_OFFSET);
+		}
+
+		item = container_of(found, struct provider_parameter, entry);
+		free(item->name);
+		free(item);
+	}
+
+	slist_insert_tail(&handle->entry, &provider_handles);
+
+	return handle;
+
+cleanup:
+	if (handle) {
+		free(handle);
+	}
+err:
+	fprintf(stderr, "%s: %s", __func__, strerror(errno));
+	return NULL;
+}
+
+/*
+ * Tokenize string from environment variable and run action function on each
+ * token found.
+ */
+static void find_options(char *input, found_func_t *action)
+{
+	char *save_ptr;
+	char *token;
+
+	if (!input) {
+		return;
+	}
+
+	for (token = strtok_r(input, ",", &save_ptr); token;
```

Comment:
I'm not overly concerned.  :)  Similar to strcasecmp, other platforms could implement it, if necessary.  It's really only an issue if other platforms define it, but with a different behavior.